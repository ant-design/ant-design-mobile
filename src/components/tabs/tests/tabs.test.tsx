import React, { useState } from 'react'
import { fireEvent, render, testA11y } from 'testing'
import Tabs, { TabsProps } from '..'

const classPrefix = `adm-tabs`

describe('Tabs', () => {
  const Basic = (props: TabsProps) => (
    <Tabs {...props}>
      <Tabs.Tab title='fruits' key='fruits'>
        Apple
      </Tabs.Tab>
      <Tabs.Tab title='vegetables' key='vegetables'>
        Tomato
      </Tabs.Tab>
      <Tabs.Tab title='animals' key='animals'>
        Ant
      </Tabs.Tab>
    </Tabs>
  )
  test('a11y', async () => {
    await testA11y(<Basic />)
  })

  test('basic usage', async () => {
    const { getByText } = render(<Basic defaultActiveKey='animals' />)
    expect(getByText('Ant')).toBeVisible()
    expect(getByText('animals')).toHaveClass(`${classPrefix}-tab-active`)
    fireEvent.click(getByText('vegetables'))
    expect(getByText('Tomato')).toBeVisible()
    expect(getByText('vegetables')).toHaveClass(`${classPrefix}-tab-active`)
  })

  test('controlled mode', async () => {
    const App = () => {
      const [activeKey, setActiveKey] = useState<string | null>(null)
      return <Basic activeKey={activeKey} onChange={key => setActiveKey(key)} />
    }
    const { getByText } = render(<App />)
    expect(document.querySelectorAll(`.${classPrefix}-tab-active`).length).toBe(
      0
    )
    fireEvent.click(getByText('vegetables'))
    expect(getByText('Tomato')).toBeVisible()
  })

  test('disabled tab', async () => {
    const onChange = jest.fn()
    const { getByText } = render(
      <Tabs onChange={onChange}>
        <Tabs.Tab title='fruits' key='fruits' />
        <Tabs.Tab title='vegetables' key='vegetables' />
        <Tabs.Tab title='animals' key='animals' disabled />
      </Tabs>
    )

    expect(getByText('animals')).toHaveClass(`${classPrefix}-tab-disabled`)
    fireEvent.click(getByText('animals'))
    expect(onChange).not.toBeCalled()
  })

  test('`activeLineMode` prop', async () => {
    const { container: fullContainer } = render(<Basic activeLineMode='full' />)
    const { container: fixedContainer } = render(
      <Basic activeLineMode='fixed' />
    )
    expect(fullContainer).toMatchSnapshot()
    expect(fixedContainer).toMatchSnapshot()
  })

  test('not stretch', async () => {
    const { getByTestId } = render(<Basic stretch={false} data-testid='tabs' />)
    expect(getByTestId('tabs')).not.toHaveClass(
      `${classPrefix}-tab-wrapper-stretch`
    )
  })

  test('render the DOM structure when hidden', async () => {
    const { queryByText } = render(
      <Tabs>
        <Tabs.Tab title='fruits' key='fruits'>
          Apple
        </Tabs.Tab>
        <Tabs.Tab title='vegetables' key='vegetables'>
          Tomato
        </Tabs.Tab>
        <Tabs.Tab title='animals' key='animals' forceRender>
          Ant
        </Tabs.Tab>
      </Tabs>
    )

    expect(queryByText('Ant')).toBeInTheDocument()
  })

  test('unmount content when not visible', async () => {
    const { getByText, queryByText } = render(
      <Tabs>
        <Tabs.Tab title='fruits' key='fruits' destroyOnClose>
          Apple
        </Tabs.Tab>
        <Tabs.Tab title='vegetables' key='vegetables'>
          Tomato
        </Tabs.Tab>
        <Tabs.Tab title='animals' key='animals'>
          Ant
        </Tabs.Tab>
      </Tabs>
    )
    expect(queryByText('Apple')).toBeInTheDocument()
    fireEvent.click(getByText('vegetables'))
    expect(queryByText('Apple')).not.toBeInTheDocument()
  })

  test('RTL render component should be rendered correctly in RTL direction', async () => {
    const { container } = render(
      <Tabs direction='rtl'>
        <Tabs.Tab title='fruits' key='fruits'>
          Apple
        </Tabs.Tab>
        <Tabs.Tab title='vegetables' key='vegetables'>
          Tomato
        </Tabs.Tab>
      </Tabs>
    )
    expect(container).toMatchSnapshot()
  })

  test('tabs should be focusable and keyboard-navigable', async () => {
    const { getByText } = render(<Basic />)

    const fruitsTab = getByText('fruits')
    const vegetablesTab = getByText('vegetables')
    const animalsTab = getByText('animals')

    expect(fruitsTab).toHaveAttribute('tabIndex', '0')
    expect(vegetablesTab).toHaveAttribute('tabIndex', '-1')
    expect(animalsTab).toHaveAttribute('tabIndex', '-1')

    fruitsTab.focus()
    expect(fruitsTab).toHaveFocus()

    fireEvent.keyDown(fruitsTab, { key: 'ArrowRight' })
    expect(vegetablesTab).toHaveFocus()
    expect(vegetablesTab).toHaveClass(`${classPrefix}-tab-active`)
    expect(fruitsTab).toHaveAttribute('tabIndex', '-1')
    expect(vegetablesTab).toHaveAttribute('tabIndex', '0')

    fireEvent.keyDown(vegetablesTab, { key: 'ArrowRight' })
    expect(animalsTab).toHaveFocus()
    expect(animalsTab).toHaveClass(`${classPrefix}-tab-active`)

    fireEvent.keyDown(animalsTab, { key: 'ArrowLeft' })
    expect(vegetablesTab).toHaveFocus()
    expect(vegetablesTab).toHaveClass(`${classPrefix}-tab-active`)
  })
})
