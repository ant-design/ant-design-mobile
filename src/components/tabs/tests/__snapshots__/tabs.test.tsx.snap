// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tabs \`activeLineMode\` prop 1`] = `
<div>
  <div
    class="adm-tabs"
    style="direction: ltr;"
  >
    <div
      class="adm-tabs-header"
    >
      <div
        class="adm-tabs-header-mask adm-tabs-header-mask-left"
        style="opacity: 0;"
      />
      <div
        class="adm-tabs-header-mask adm-tabs-header-mask-right"
        style="opacity: 0;"
      />
      <div
        class="adm-tabs-tab-list"
        role="tablist"
      >
        <div
          class="adm-tabs-tab-line"
          style="width: 0px; transform: none;"
        />
        <div
          class="adm-tabs-tab-wrapper adm-tabs-tab-wrapper-stretch"
        >
          <div
            aria-selected="true"
            class="adm-tabs-tab adm-tabs-tab-active"
            role="tab"
            tabindex="0"
          >
            fruits
          </div>
        </div>
        <div
          class="adm-tabs-tab-wrapper adm-tabs-tab-wrapper-stretch"
        >
          <div
            aria-selected="false"
            class="adm-tabs-tab"
            role="tab"
            tabindex="-1"
          >
            vegetables
          </div>
        </div>
        <div
          class="adm-tabs-tab-wrapper adm-tabs-tab-wrapper-stretch"
        >
          <div
            aria-selected="false"
            class="adm-tabs-tab"
            role="tab"
            tabindex="-1"
          >
            animals
          </div>
        </div>
      </div>
    </div>
    <div
      class="adm-tabs-content"
      style="display: block;"
    >
      Apple
    </div>
  </div>
</div>
`;

exports[`Tabs \`activeLineMode\` prop 2`] = `
<div>
  <div
    class="adm-tabs"
    style="direction: ltr;"
  >
    <div
      class="adm-tabs-header"
    >
      <div
        class="adm-tabs-header-mask adm-tabs-header-mask-left"
        style="opacity: 0;"
      />
      <div
        class="adm-tabs-header-mask adm-tabs-header-mask-right"
        style="opacity: 0;"
      />
      <div
        class="adm-tabs-tab-list"
        role="tablist"
      >
        <div
          class="adm-tabs-tab-line"
          style="transform: none;"
        />
        <div
          class="adm-tabs-tab-wrapper adm-tabs-tab-wrapper-stretch"
        >
          <div
            aria-selected="true"
            class="adm-tabs-tab adm-tabs-tab-active"
            role="tab"
            tabindex="0"
          >
            fruits
          </div>
        </div>
        <div
          class="adm-tabs-tab-wrapper adm-tabs-tab-wrapper-stretch"
        >
          <div
            aria-selected="false"
            class="adm-tabs-tab"
            role="tab"
            tabindex="-1"
          >
            vegetables
          </div>
        </div>
        <div
          class="adm-tabs-tab-wrapper adm-tabs-tab-wrapper-stretch"
        >
          <div
            aria-selected="false"
            class="adm-tabs-tab"
            role="tab"
            tabindex="-1"
          >
            animals
          </div>
        </div>
      </div>
    </div>
    <div
      class="adm-tabs-content"
      style="display: block;"
    >
      Apple
    </div>
  </div>
</div>
`;

exports[`Tabs RTL render component should be rendered correctly in RTL direction 1`] = `
<div>
  <div
    class="adm-tabs"
    style="direction: rtl;"
  >
    <div
      class="adm-tabs-header"
    >
      <div
        class="adm-tabs-header-mask adm-tabs-header-mask-left"
        style="opacity: 0;"
      />
      <div
        class="adm-tabs-header-mask adm-tabs-header-mask-right"
        style="opacity: 0;"
      />
      <div
        class="adm-tabs-tab-list"
        role="tablist"
      >
        <div
          class="adm-tabs-tab-line"
          style="width: 0px; transform: none;"
        />
        <div
          class="adm-tabs-tab-wrapper adm-tabs-tab-wrapper-stretch"
        >
          <div
            aria-selected="true"
            class="adm-tabs-tab adm-tabs-tab-active"
            role="tab"
            tabindex="0"
          >
            fruits
          </div>
        </div>
        <div
          class="adm-tabs-tab-wrapper adm-tabs-tab-wrapper-stretch"
        >
          <div
            aria-selected="false"
            class="adm-tabs-tab"
            role="tab"
            tabindex="-1"
          >
            vegetables
          </div>
        </div>
      </div>
    </div>
    <div
      class="adm-tabs-content"
      style="display: block;"
    >
      Apple
    </div>
  </div>
</div>
`;
